### PR TITLE
Add a few transform convenience methods

### PR DESCRIPTION
**Pull request by Benjamin-L**
Converting transforms into a matrix in order to operate on vectors or compute the inverse is poor ergonomics, and these operations can be implemented easily as methods in `Transform` and `GlobalTransform`. I think the name `transform_vector` is a bit confusing, especially given `mul_vec3` exists. If anybody has a better suggestion that would be great!

### CODE REVIEW

File: crates/bevy_transform/src/components/global_transform.rs [hunk 1]
```diff
@@ -195,6 +211,24 @@ impl GlobalTransform {
         value
     }
 
+    /// Returns the result of applying this [`GlobalTransform`] to a [`Vec3`] interpreted as a point.
+    ///
+    /// This applies rotation, scale, and translation. It's the equivalent of using w=1 in
+    /// homogeneous coordinates.
+    #[inline]
+    pub fn transform_point(&self, point: Vec3) -> Vec3 {
+        self.scale * (self.rotation * point)
```
Comment (alice-i-cecile) [inline crates/bevy_transform/src/components/global_transform.rs:220]: How does translation get applied here?

File: crates/bevy_transform/src/components/global_transform.rs [hunk 2]
```diff
@@ -195,6 +211,24 @@ impl GlobalTransform {
...
+    }
+
+    /// Returns the result of applying this [`GlobalTransform`] to a [`Vec3`] interpreted as a vector.
+    ///
+    /// This applies rotation and scale, but not translation. It's the equivalent of using w=0 in
```
Comment (alice-i-cecile) [inline crates/bevy_transform/src/components/global_transform.rs:225]: Did we swap the comments / math here?

File: crates/bevy_transform/src/components/global_transform.rs [hunk 3]
```diff
@@ -195,6 +211,24 @@ impl GlobalTransform {
         value
     }
 
+    /// Returns the result of applying this [`GlobalTransform`] to a [`Vec3`] interpreted as a point.
+    ///
+    /// This applies rotation, scale, and translation. It's the equivalent of using w=1 in
+    /// homogeneous coordinates.
```
Comment (alice-i-cecile) [inline crates/bevy_transform/src/components/global_transform.rs:217]: Could you link an appropriate explanatory article on homogenous coordinates as part of this? It's pretty dense for beginners.

https://www.tomdalling.com/blog/modern-opengl/explaining-homogenous-coordinates-and-projective-geometry/ looks quite nice.

File: crates/bevy_transform/src/components/transform.rs [hunk 4]
```diff
@@ -124,6 +124,22 @@ impl Transform {
         Mat4::from_scale_rotation_translation(self.scale, self.rotation, self.translation)
     }
 
+    /// Returns the inverse of this transform.
+    ///
+    /// The inverse consists of applying the inverse rotation, scale, and
+    /// translation in reverse order.
```
Comment (alice-i-cecile) [inline crates/bevy_transform/src/components/transform.rs:130]: I think spelling out A, then B, then C is clearer than "reverse order".

File: crates/bevy_transform/src/components/transform.rs [hunk 5]
```diff
@@ -207,6 +223,24 @@ impl Transform {
         value
     }
 
+    /// Returns the result of applying this [`Transform`] to a [`Vec3`] interpreted as a point.
+    ///
+    /// This applies rotation, scale, and translation. It's the equivalent of using w=1 in
+    /// homogeneous coordinates.
+    #[inline]
+    pub fn transform_point(&self, point: Vec3) -> Vec3 {
```
Comment (alice-i-cecile) [inline crates/bevy_transform/src/components/transform.rs:231]: Does having two functions of the same name but different function arguments work in Rust?

Comment (NathanSWard) [reply crates/bevy_transform/src/components/transform.rs:231]: rust doesn't have function overloading, so it should not work.

**Comment by NathanSWard** [comment]
I would like to see some tests added to ensue these functions have the correct behavior :smile:

File: crates/bevy_transform/src/components/global_transform.rs (hunk 2 above)
Comment (Benjamin-L) [reply crates/bevy_transform/src/components/global_transform.rs:225]: Yeah... I sure did. lol. I'll fix that

**Comment by Benjamin-L** [comment]
Is there a preference between doc-tests vs. putting the tests in their own functions? I'm gonna add doc-tests for now, but I can change that if you want me to.

File: crates/bevy_transform/src/components/transform.rs (hunk 5 above)
Comment (bjorn3) [reply crates/bevy_transform/src/components/transform.rs:231]: One of the functions is on `Transform`, the other on `GlobalTransform`.

**Comment by mockersf** [comment]
doc tests are when the test is useful to show usage in the doc. if it's actual tests, they should live in `#[test]` instead so that they won't be visible in the doc

**Comment by Benjamin-L** [comment]
While writing unit tests I discovered that the existing matrix methods are wrong. The `Mat4::from_scale_rotation_translation` function applies scale first, then rotation, then translation. The bevy `Transform` applies rotation first, then scale, then translation. The result is that the behavior of matrices from the `Transform::compute_matrix` method are not consistent with the transform they were generated from.

At this point we have a choice. We can either modify `Transform::compute_matrix` to use the RST (rotate, scale, translate) ordering, or we can modify the entire `Transform` type to use SRT. I'm actually in favor of the latter, because SRT is *far* more common, and I personally have run into several errors in my code using bevy because I assumed that scale was applied before rotation. Either option is a breaking change, but modifying the actual behavior of the `Transform` struct would probably be more significant. I would be in favor of doing this as soon as possible, while there aren't as many users. It's also worth noting that the inconsistency only matters with nonuniform scale. If scale is uniform the rotation and translation operations are commutative.

**Comment by alice-i-cecile** [comment]
I heavily prefer SRT as well and agree with your reasoning. If this change is implemented, please note it in the [migration guide](https://github.com/bevyengine/bevy-website/pull/134) due to the serious impact on users.

**Comment by alice-i-cecile** [comment]
(also sorry @cart, CI approval on this one would be nice too)

**Comment by Benjamin-L** [comment]
I could make a separate PR to switch to SRT, or I could make it part of this one. If we do a separate PR I'll probably wait for it to get merged before pushing the tests for this, since they depend on the transform and matrix behavior being consistent.

**Comment by alice-i-cecile** [comment]
@Benjamin-L, I think a separate PR is better: it'll keep the scope smaller for review and keep this PR uncontroversial.

**Comment by Benjamin-L** [comment]
Alright, I'll mark this one as WIP then since it won't be complete until the other one is done.

**Comment by Benjamin-L** [comment]
Another discovery: the transform composition method is also broken, even for uniform scaling. I'm gonna use RST ordering, since that's what the code currently does, but this applies for SRT as well.

```
t⃗ = Translation component (a vector)
s⃗ = scale component (a vector, multiplied component-wise)
R = rotation component (a matrix, it's a quaternion in the real implementation but whatever)
```
For a transform `tₙ(x)`:
```
tₙ(x) = Rₙ s⃗ₙ x + t⃗ₙ
```
The correct way to compose transformations looks like this:
```
t₂(x) = t₁(t₀(x))
t₂(x) = R₁ s⃗₁ (R₀ s⃗₀ x + t⃗₀) + t⃗₁
      = R₁ s⃗₁ R₀ s⃗₀ x + R₁ s⃗₁  t⃗₀ + t⃗₁
      = (R₁ R₀) ((R₀⁻¹ s⃗₁) s⃗₀) x + R₁ s⃗₁  t⃗₀ + t⃗₁
R₂ = R₁ R₀
s⃗₂ = (R₀⁻¹ s⃗₁) s⃗₀
t⃗₂ = R₁ s⃗₁  t⃗₀ + t⃗₁
```
What the current code does is:
```
R₂ = R₁ R₀
s⃗₂ = s⃗₁ s⃗₀
t⃗₂ = t⃗₀ + t⃗₁
```
This needs to get fixed regardless of how we resolve the discrepancy between `transform.compute_matrix() * vec` and `transform * vec`. I think it probably belongs in a third PR that will be less controversial than the SRT vs RST change, but still controversial because it is a subtle breaking change.

**Comment by Benjamin-L** [comment]
Actually, there's also an argument for doing both of those changes as part of the same PR: I want to add a bunch of unit tests that will catch this type of thing in the future. The unit tests won't pass until both problems are fixed.

**Comment by alice-i-cecile** [comment]
Nice find! Thanks a ton. I like both of those fixes in the same PR so then our unit tests can pass.

**Comment by Benjamin-L** [comment]
What are your thoughts on renaming the `mul_vec3` function to `transform_point` in this PR, instead of keeping both around. If we're gonna be making breaking changes before the next version anyway, I think the name `transform_vec3` is confusing and does a bad job of communicating to users that they need to consider whether the `Vec3` they are transforming is a point or a vector.

**Comment by alice-i-cecile** [comment]
> What are your thoughts on renaming the mul_vec3 function to transform_point in this PR, instead of keeping both around.

I think this is a good change. I much prefer focusing on use cases rather than the raw math, and we should try to avoid redundant approaches like this where possible. Adding a doc alias would help with migration pain though :)

**Comment by Benjamin-L** [comment]
Deprecating the old one instead of removing it would also be an option for less migration pain.

**Comment by alice-i-cecile** [comment]
> Deprecating the old one instead of removing it would also be an option for less migration pain.

Ha, I think deprecating functionality violates our Bevy ethos at this stage ;) I'd rather have pointers to the fix than worry about managing deprecation at this stage, especially since nothing else in the engine uses a deprecation strategy yet.

(You're fully right that this is the right solution for more mature libraries)

**Comment by Benjamin-L** [comment]
> I much prefer focusing on use cases rather than the raw math

Honestly, if I was designing this system from scratch I would make the distinction between points and vectors at the type level. I've done this in most of the graphics-related things I've worked on and it has caught a ton of bugs. It does come at the cost of some increased API friction when you do need to convert between the two though, and it might be too late to consider it at this point.

**Comment by alice-i-cecile** [comment]
> > I much prefer focusing on use cases rather than the raw math
> 
> Honestly, if I was designing this system from scratch I would make the distinction between points and vectors at the type level. I've done this in most of the graphics-related things I've worked on and it has caught a ton of bugs. It does come at the cost of some increased API friction when you do need to convert between the two though, and it might be too late to consider it at this point.

Have you seen @aevyrie's [geometric primitives RFC](https://github.com/bevyengine/rfcs/pull/12). This sort of thing is exactly the intent of that work, and would make an excellent comment to include there.

**Comment by Benjamin-L** [comment]
Ooh, I did not see that RFC before. It's pretty much exactly what I had in mind.

**Comment by alice-i-cecile** [comment]
Once that RFC is implemented, would it make sense to move some of these convenience methods to those structs or traits instead? It makes more sense to me than them being methods of the `Transform` itself TBH, just because I tend to think of methods as transforming the object itself, rather than applying the object to something else.

**Comment by Benjamin-L** [comment]
I think having methods that perform an action on a different object in the context of the `self` object is pretty normal, but I do think we should add `Mul<Point>` and `Mul<Direction>` impls. I guess I will also say that if I was looking for the methods to apply a transform to a point or direction without prior knowledge about the API, the `Transform` object is probably the first place I would look.

**Comment by alice-i-cecile** [comment]
The Mul impls seem like a nice compromise; that sounds good to me.

**Comment by Benjamin-L** [comment]
Hmm, maybe it makes sense to have a more general `Transformable` trait. I just realized that it would make sense to apply a transform to several of the other types added in the RFC. Going even farther with this, maybe we want to make a distinction between isometries, similiarities, and affine transformations. You can apply an isometry or similarity to a circle or arc, and the result will still be a circle/arc, but you can't do the same with an arbitrary affine transformation because of the nonuniform scale.

**Comment by alice-i-cecile** [comment]
> Hmm, maybe it makes sense to have a more general Transformable trait

I think I agree with you, but this deserve to be a comment on the RFC linked or an RFC of its own :p

**Comment by Benjamin-L** [comment]
So... I just realized something. There's a reason that the order of application for `Transform` is RST.  Then SRT representation is actually not closed under composition, because the result of composing arbitrary nonuniform with nonuniform scale can have shear, which cannot be captured by the SRT representation. This makes me a little sad, because RST is kinda unintuitive, but it does make the decision of how to resolve the discrepancy between applying a transform directly vs converting it to a matrix easy, because we only have one option.

**Comment by alice-i-cecile** [comment]
> There's a reason that the order of application for Transform is RST. SRT representation is actually not closed under composition, because the result of composing arbitrary transforms with nonuniform scale can have shear, which cannot be captured by the SRT representation. 

Ah, you're right. Well, that makes the decision easy. We should try to document both the ordering and why it's needed carefully then.

**Comment by mtsr** [comment]
Some of the math errors would also be solved by #1755, I think.

**Comment by Benjamin-L** [comment]
Oh, yes they would... It might be a good idea to wait on a decision on that issue before continuing with this.

**Comment by cart** [comment]
Yeah I'm cool with making the relevant changes in a single pr if it means adding tests for general transform op correctness.

**Comment by cart** [comment]
Wider gamedev ecosystem Transform survey: https://github.com/bevyengine/bevy/issues/1755#issuecomment-832917834

I think we should stick with the current SRT impl

**Comment by alice-i-cecile** [comment]
@Benjamin-L, can you comment in #2373? We've since relicensed to MIT + Apache, and I'd like to either ensure that this can be picked up or close it out.

**Comment by alice-i-cecile** [comment]
Closing this out as abandoned. Do not reuse this code unless the relicensing is resolved.

**Pull request closed without merging**